pkg/controller/certificate-shim/helper.go

```diff
@@ -110,5 +110,84 @@ func translateAnnotations(crt *cmapi.Certificate, ingLikeAnnotations map[string]
 		crt.Spec.RevisionHistoryLimit = pointer.Int32(int32(limit))
 	}
 
+	if privateKeyAlgorithm, found := ingLikeAnnotations[cmapi.PrivateKeyAlgorithmAnnotationKey]; found {
+		algorithm := cmapi.PrivateKeyAlgorithm(privateKeyAlgorithm)
+		switch algorithm {
+		case cmapi.RSAKeyAlgorithm,
+			cmapi.ECDSAKeyAlgorithm,
+			cmapi.Ed25519KeyAlgorithm:
+			// ok
+		default:
+			return fmt.Errorf("%w %q: invalid private key algorithm %q", errInvalidIngressAnnotation, cmapi.PrivateKeyAlgorithmAnnotationKey, privateKeyAlgorithm)
+		}
+
+		if crt.Spec.PrivateKey == nil {
+			crt.Spec.PrivateKey = &cmapi.CertificatePrivateKey{Algorithm: algorithm}
+		} else {
+			crt.Spec.PrivateKey.Algorithm = algorithm
+		}
+	}
+
+	if privateKeyEncoding, found := ingLikeAnnotations[cmapi.PrivateKeyEncodingAnnotationKey]; found {
+		encoding := cmapi.PrivateKeyEncoding(privateKeyEncoding)
+		if encoding != cmapi.PKCS1 &&
+			encoding != cmapi.PKCS8 {
+			return fmt.Errorf("%w %q: invalid private key encoding %q", errInvalidIngressAnnotation, cmapi.PrivateKeyEncodingAnnotationKey, privateKeyEncoding)
+		}
+
+		if crt.Spec.PrivateKey == nil {
+			crt.Spec.PrivateKey = &cmapi.CertificatePrivateKey{Encoding: encoding}
+		} else {
+			crt.Spec.PrivateKey.Encoding = encoding
+		}
+	}
+
+	if privateKeySize, found := ingLikeAnnotations[cmapi.PrivateKeySizeAnnotationKey]; found {
+		size, err := strconv.Atoi(privateKeySize)
+		if err != nil {
+			return fmt.Errorf("%w %q: %v", errInvalidIngressAnnotation, cmapi.PrivateKeySizeAnnotationKey, err)
+		}
+
+		// default algorithm
+		algorithm := cmapi.RSAKeyAlgorithm
+		if crt.Spec.PrivateKey != nil && crt.Spec.PrivateKey.Algorithm != "" {
+			algorithm = crt.Spec.PrivateKey.Algorithm
+		}
+
+		switch algorithm {
+		case cmapi.RSAKeyAlgorithm:
+			if size < 2048 {
```

Comment:
Yeah, I intentionally made the validation stricter because I think of the annotations as an override mechanism.
So I would want to get an error if I override with a value that is actually ignored.

Reply:
> We also want to error out if the size > 8192 (see [webhook validation for `certificate.spec.privateKey`](https://github.com/cert-manager/cert-manager/blob/master/internal/apis/certmanager/validation/certificate.go#L115)

Thanks, fixed it